semgrep-core/src/tainting/Dataflow_tainting.ml

```diff
@@ -581,9 +589,12 @@ and check_tainted_lval_offset env offset =
   | Index e ->
       let taints, lval_env = check_tainted_expr env e in
       let taints =
-        if not env.options.taint_assume_safe_indexes then taints
-        else (* Taint from the index should be ignored. *)
+        if
+          env.options.taint_assume_safe_indexes
+          || not env.options.taint_builtin_propagators
```

Comment:
This is getting complicated to read. I could easily imagine myself adding a `assume_safe_fields` but forgetting to add `taint_builtin_propagators`. 

I would prefer that we have one central place where we read the taint options and then set the fields `propagate_through_default_propagators`, `propagate_through_indexes`, and `propagate_through_functions`. Then, when deciding whether or not to return taints, we only have to check one thing.

The central location makes it easy to reason about how to set each field (because you have all the context of all the other fields) and each if statement is a clear and consistent check